### ISSUE TITLE
Receiving Bad Request when calling MINIO Server from minio/api.py (

### ISSUE THREAD

**Issue by lmcdasm** [minio/minio-py#947]
Hello.

I am working with @vsoch  on an SREGISTRY integration that uses MINIO for the storage backend with HTTPS/SSL enabled.

The environment (for testing/development) is a couple containers on a single host and looks like this:

Where the GREEN Container is our running instance of MINIO and the Two containers in purple are our MINIO clients that implement / instantiate two instances of MinioClient on startup _(as i will outline below)_

![image](https://user-images.githubusercontent.com/9358276/88952485-83069d00-d265-11ea-81a8-bdb5db7df382.png)

A bit about our minio installation

Here is the docker-compose that is in use - you can see we consuming the .mino-env  on startup
- we also have done the appropriate SSL certificate (with the .PEM consisting our of cert+intermediate and the key the key)
-  "know" this works, since we can see the following when we do a a simple "curl" from linux (since a browser test might be helped by baked in intermediates) we get the following (i have redacted some of our personal info)

![image](https://user-images.githubusercontent.com/9358276/88961811-89037a80-d273-11ea-8852-b534a2fa8e61.png)

![image](https://user-images.githubusercontent.com/9358276/88953340-cad9f400-d266-11ea-8990-983877fdcb5b.png)

When we login to the MINIO Web UI, we can see the following:
 - we see that SSL/HTTP is working (Browser wise)  (UI-pic1 below)
 - we are able to see that we can create new Containers and they are written back down the backing filesystem (UI-pic2-1, UI-PIC2-2,  and TERM-SHOT-1 of the filesystem after i created)

UI-PIC-1
![image](https://user-images.githubusercontent.com/9358276/88953895-8c910480-d267-11ea-969e-2fac87896cc3.png)

UI-PIC-2-1
![image](https://user-images.githubusercontent.com/9358276/88953623-302de500-d267-11ea-942a-acb34ed13def.png)

UI-PIC-2-2
![image](https://user-images.githubusercontent.com/9358276/88953747-5e132980-d267-11ea-8f76-da0d54142388.png)

TERM-SHOT-1
![image](https://user-images.githubusercontent.com/9358276/88953997-b518fe80-d267-11ea-8a87-344ba67847f7.png)

As well, here is the "mc config ls" output to show our setup defined

MC-LS-1

![image](https://user-images.githubusercontent.com/9358276/88954197-fad5c700-d267-11ea-9e6f-c6e0f2153904.png)

So we are pretty sure that the HTTPS action to/from MINIO is 'ok" from a certificate point of view.

In the SREGISTRY instantion of the two minio clients, we have the following code bit (and this is where we are seeing the "BadRequest failure "  - note that we create two clients, one for External Access - which is HTTPS://ourfdqn as shown and there is another that is to use the internal link  - minio - which would resolve on the docker "links" (we are not using a seperate brigde as docker in the docker compose above to link the client containers to the minio container).  the idea being that the sregistries are trusted and thus would not implement https.  However, it seems that with the secure=TRUE/FALSE, flag we end up in a situation where minio is bringing up a endpoint called https://127.0.0.1:9000.

[root@grid-singregistry-master sregistry]# docker logs -f 401a3d72a310
Attempting rotation of encrypted config, IAM users and policies on MinIO with newly supplied credentials
Rotation complete, please make sure to unset MINIO_ACCESS_KEY_OLD and MINIO_SECRET_KEY_OLD envs
Endpoint:  https://172.17.0.4:9000  https://127.0.0.1:9000

The issue here is that for me to "make this work" i would need to put a SAN in our cert for 127.0.0.1. 

CLIENT-CODE-SNIP-1

![image](https://user-images.githubusercontent.com/9358276/88961562-32963c00-d273-11ea-97b3-e8b98a0c150d.png)

CLIENT-STACK-TRACE-1

![image](https://user-images.githubusercontent.com/9358276/88954709-c9a9c680-d268-11ea-90a0-981bdbf8f9b5.png)

From what i have understood, is that when we instantiate the minioClient call( either the internal or the External one) there is a connection, so im a bit stumped as to why this is failed with BadRequest on the if not makebucket command.

You will notice above that the exception shows a minioExternalClient call, that is from me trying both types of clients to see if there is anything instantiated.

I did a TCPDUMP on port 9000 on the NIC that hold the docker bridge and container traffic and no request on port 9000 ever arrived during any startup, so i suspect that the instantiation is failing of the client, but im not sure how to further in minio.py figure out if its cause of a SSL/non SSL mix up (our idea is that all communications shall be over HTTPS (internal or external as we start to distribute it).

Perhaps a silly configuration i am missing with the secure flag for the clients?

Any help would be much appreciated.
D

**Comment by vsoch** [comment]
This is beautifully documented @lmcdasm ! And we can also provide detail that without ssl (just http), it works without a hitch. And @lmcdasm can you please comment if you tried keeping the internal client to just use http (and if it worked?)

**Comment by lmcdasm** [comment]
Hello

@vsoch  - will try to address your points as well and illustrate.

To continue with the Thread of the current "makeBucket call that throws a bad request" and since i can create a bucket in the MINIO BACKEND, i decided to comment out that call and test it.  in this way, we can see that the components start up (since there is no "usage") of the call to the MINIO endpoint of yet.. For clarity sake, i show the same code call as above, but simply commented out.

The reason that i do this is when i then do a singularity "push" - we can see when the minio client is called that i get an exception of ACCESS/SECRET KEY mismatch - so this, in the end we (@vsoch  and I) think that the issue lies is that the make_bucket is failing cause there is a failed url call, and while its not hte URL persay, i had seen in earlier configuration attempts that if you are going to run an internal endpoint with https://127.0.0.1 then you have to embed a SAN into your certificate - which i could do, but not really ideal in production based environments.

I put this here for illustration - the initial call on the internal client (which is a check both for connectivity and the existence of the bucket) is a good one and "should" by all rights work to my eyes, however i think either i have to have a fqdn (internal versus external) on the internal minio endpoint in docker layout if its going to work.

At the same time, i tried with doing the "make_bucket" on the external client - which should have work, and that throws the same error as well

CODE-SNIP-COMMENT-OUT-MAKE-BUCKET

![image](https://user-images.githubusercontent.com/9358276/88967452-d8e63f80-d27b-11ea-9306-a1e9181b1b09.png)


Thus from when we run our "singularity push" - which hits our sregisty app and then minio essentially we can see we get to the point of the pushing.

![image](https://user-images.githubusercontent.com/9358276/88967614-1054ec00-d27c-11ea-8cde-22a61ad4994c.png)


and in our client to minio log, we can see that instead of Badrequest, we throw a ACCESS/SECRET key error

![image](https://user-images.githubusercontent.com/9358276/88967752-39757c80-d27c-11ea-84ba-d6a659bfaac3.png)


this seems somehow related to something i saw with MC and the fact that you do a Timer on a SIGNED request somewhere internally? i could be way off base, but during my diggings, it seems a bit strange - again, i never see any traffic arrive at port 9000 

now - to @vsoch  request, I will roll the HTTP (no ssl) equivalent setup and post for comparisons and hopefully that will help out. - give me a hour or so to roll it out and ill re-post the same bits.

Cheers,
D

**Comment by vsoch** [comment]
@lmcdasm the call that is failing is not an interaction between the external client and server (although your push originally triggers it) but rather it is the internal client requesting to generate a signed url (to return to the external client). This call fails, which is why you don't see any traffic at port 9000.

**Comment by lmcdasm** [comment]
> @lmcdasm the call that is failing is not an interaction between the external client and server (although your push originally triggers it) but rather it is the internal client requesting to generate a signed url (to return to the external client). This call fails, which is why you don't see any traffic at port 9000.

excellent - that could the thread to pull on here.. as i said, in some Minio documentation that there is a part to set an unsigned call  - however when i do the mc add part and provide the key (myminio) i thought this would do it..

**Comment by vsoch** [comment]
@lmcdasm the url is generated [here](https://github.com/singularityhub/sregistry/blob/89caaf05f91f63fa34e3f461eb469603d9f7ddcc/shub/apps/library/views/minio.py#L94) and note that there is an argument for any extra headers that are needed. I think a solid way to start debugging would be to generate this URL for a working call (with http) and then one with https (not working). 

 - [the multipart upload is originally requested here](https://github.com/singularityhub/sregistry/blob/89caaf05f91f63fa34e3f461eb469603d9f7ddcc/shub/apps/library/views/images.py#L293) (this generates an upload id we can return to the external client
 - [then the PUT request generates the signed URL just for each blob here](https://github.com/singularityhub/sregistry/blob/89caaf05f91f63fa34e3f461eb469603d9f7ddcc/shub/apps/library/views/images.py#L243)

So basically: we POST from the external client to the server to request the multipart. Then the internal client starts the request to generate an upload id (and other metadata) that is returned to the external client. The external client then uses the upload ID for each chunk, doing a PUT to request the URL where to put the chunk. So likely the current setup is failing at the original request (the one to generate the upload id) which comes down to:
```python
res = s3.create_multipart_upload(Bucket=MINIO_BUCKET, Key=storage)
```

**Comment by vsoch** [comment]
And keep trying @lmcdasm - you're doing great, detailed work, and figuring this out is going to be hugely useful for the community! Don't worry if it takes some time, it took me ~1.5 months to get the original implementation working.

**Comment by lmcdasm** [comment]
> And keep trying @lmcdasm - you're doing great, detailed work, and figuring this out is going to be hugely useful for the community! Don't worry if it takes some time, it took me ~1.5 months to get the original implementation working.

Thanks.. the one thing in the end is the docker vs. the long term K8s deployment play.. in our world, we have ingress controllers (that will take the place of nginx) and we will use a service mesh in k8s for certificate management..  in that sense, we wouldnt have a "internal and external client" - there would be "one call" to any publish endpoint (since they are all backed by a HTTPS ingress controller that is cluster-wide">.

i did notice that there is a context switch.. int he make_bucket at the start, you are using the minioClient (internal), however for the presign call you are switch back out to the HTTPS,( i tried the makebucket with the ExternalClient object as well and got the same result), but it was something i noticed, perhaps we need to add a check at instantiation time that both endpoints are up and working?

will keep at it.. had some family time and will post the http flavour shortly.

**Comment by vsoch** [comment]
I didn't stop / exit from the server in case it didn't work because it would be shown in the logs and subsequent failure, and also it would make it harder to debug if I exit immediately. But I definitely agree maybe there could be some way to better show that.

**Comment by lmcdasm** [comment]
One thing i did notice as well (from the external client side, i think we are getting as far as URL generation, since i see that the scs-library (Singuliary) apiGet - they are getting the hash that you generate and attempting to make the post to that URL

**postFileV2 below**

DEBUG   [U=0,P=60094]      apiGet()                      apiGet calling v1/entities/daniel.smith
DEBUG   [U=0,P=60094]      apiGet()                      apiGet calling v1/collections/daniel.smith/dasm-test
DEBUG   [U=0,P=60094]      apiGet()                      apiGet calling v1/containers/daniel.smith/dasm-test/something

DASM This is i assume the registry given the hash for the container to be uploaded
**DEBUG   [U=0,P=60094]      apiGet()                      apiGet calling v1/images/daniel.smith/dasm-test/something:sha256.b226c379e19853f16f1dc76b13b70deaf822afd6078c5bc3cd9b921d32cf7101?arch=amd64**
DEBUG   [U=0,P=60094]      postFileWrapper()             Now uploading to the library
DEBUG   [U=0,P=60094]      postFileV2()                  Using legacy (single part) uploader
DASM - here singularity is using their postMethod to that ULL
DEBUG   [U=0,P=60094]      legacyPostFileV2()            legacyPostFileV2 calling v2/imagefile/15
DASM - here they try to post calling the endpoint that was provided back - notice that it uses and endpoint of v2/imagefile/15
is that an obfuscation back to the daniel.smith/dasm-test/something:sha256  ?

Thanks alot for all the input so far

**Comment by vsoch** [comment]
I don't see the PUT request there - I see a request that goes to the legacy uploader, which isn't multipart (and is not Minio).

**Comment by lmcdasm** [comment]
> I don't see the PUT request there - I see a request that goes to the legacy uploader, which isn't multipart (and is not Minio).

hmm. im not sure i understand.. this is singularity in DEBUG calling.> So its configured to use SREGISTRY as the remote.

We can see in the UWSGI logs the following  
SINGULARITY CALLS:
DEBUG [U=0,P=60094] apiGet() apiGet calling v1/entities/daniel.smith

This INVOKES iN REGISTRY the following method 
GET NamedEntityView
which replies back with /v1/entities/daniel.smith


then we see that SINGULARITY (this is the SCS-LIBRARY that makes up the SIngularity Container Storage call as part of singularity) send the next apiGet() call 
DEBUG   [U=0,P=82236]      apiGet()                      apiGet calling v1/collections/daniel.smith/dasm-test

Which maps to SREG call of 
GET GetNamedCollectionView

then we see that SINGUARLITY CLIENT soes the 
 apiGet calling v1/containers/daniel.smith/dasm-test/something

And that maps to 
GET GetNamedContainerView

We see then that SINGUARLIT getApi() does the following call
   apiGet calling v1/images/daniel.smith/dasm-test/something:sha256.b226c379e19853f16f1dc76b13b70deaf822afd6078c5bc3cd9b921d32cf7101?arch=amd64

in the SREG logs, this generates the folowing method
GET PushNamedContainerView

Then we see the SINGULARITY try to push it:

DEBUG   [U=0,P=82236]      postFileWrapper()             Now uploading to the library
DEBUG   [U=0,P=82236]      postFileV2()                  Using legacy (single part) uploader
DEBUG   [U=0,P=82236]      legacyPostFileV2()            legacyPostFileV2 calling v2/imagefile/17
DEBUG   [U=0,P=82236]      apiCreate()                   apiCreate calling v2/imagefile/17

and this can only have been provided from the apiGet() calls above, in the REG log this method is called

[pid: 59|app: 0|req: 5/7] 10.179.111.4 () {34 vars in 453 bytes} [Thu Jul 30 16:47:51 2020] GET /version => generated 58 bytes in 7 msecs (HTTP/1.1 200) 5 headers in 141 bytes (1 switches on core 0)
POST RequestPushImageFileView

_Thus at this point, im not sure what you mean, but you "dont see a PUT" there, and that it goes to a lergacy uploader -  from the external client point of view - its using the URL that is  provided back to it? so im not sure how it (singularity) could use anything else that what is provided in the answers "back" to singularity?? This is just a simple "push" to a remote from command line, so im lost when you say "its using a legacy uploader" - reading the scs-library, it uses whatever path is provided in the previous call from the server (registry) side if i follow -

sorry if im not getting it

**Comment by vsoch** [comment]
You can take a look at the Singularity source code to see the circumstances for doing the post to `legacyPostFileV2()`: https://github.com/sylabs/scs-library-client/blob/a1fe9f14d4aaebeed059e7b5a5cbd9a88e3fdb3d/client/push.go#L266 I'll also put the code here - the endpoint is not called unless there the container is larger than the minimum chunk size:
```
func (c *Client) postFileV2(ctx context.Context, r io.ReadSeeker, fileSize int64, imageID string, callback UploadCallback, metadata map[string]string) error {
	if fileSize > minimumPartSize {
		// only attempt multipart upload if size greater than S3 minimum
		c.Logger.Log("Attempting to use multipart uploader")

		if err := c.postFileV2Multipart(ctx, r, fileSize, imageID, callback); err != nil {
			// if the error is anything other than ErrNotFound, fallback to legacy (single part)
			// uploader.
			if err != ErrNotFound {
				return err
			}
			// fallthrough to legacy (single part) uploader
		} else {
			// multipart upload successful
			return nil
		}
	}

	// fallback to legacy uploader
	c.Logger.Log("Using legacy (single part) uploader")

	return c.legacyPostFileV2(ctx, fileSize, imageID, callback, metadata)
}
```
Are you testing with a small container? You would want to enter the first loop at the top, and then either have a more clear success/fail.

**Comment by lmcdasm** [comment]
Indeed, the container is small right now, i can push to a bigger one to be more consistent, in the end i wonder though if this is "really it" since either multipart or legacy (in real terms) should work regardless.. thanks for clarifying.

**Comment by lmcdasm** [comment]
Hey there.

so i did a bit more debugging and above you mentioned you are getting the "url" fromsregistry_presign_v4 call.. however when i trace to the invoking object, i see its a call to presigned_put_object ( put some try/excepts to play around with my stub)

Code calls:

        print("DASM - storage: " + storage)
        try:
            print("DASM DEBUG - trying to get url from presigned_put_object call")
            print("DASM DEBUG - MINIO BUCKET : " + MINIO_BUCKET)
            print("DASM DEBUG - MINIO SIGNED URL EXPIRE MINUTES : " + str(MINIO_SIGNED_URL_EXPIRE_MINUTES))
            url = minioExternalClient.presigned_put_object(
               MINIO_BUCKET,
               storage,
               expires=timedelta(minutes=MINIO_SIGNED_URL_EXPIRE_MINUTES),
            )

Thus when i do my external client push, im getting this part to trigger:

Container ID: 1
DASM- Checking token
DASM - Got token
DASM - PUSH SECRET: 9ff2bdf0-2575-4769-8f38-918540c1e21c
DASM - storage: dasm-test/something:sha256.b226c379e19853f16f1dc76b13b70deaf822afd6078c5bc3cd9b921d32cf7101
DASM DEBUG - trying to get url from presigned_put_object call
DASM DEBUG - MINIO BUCKET : bucket2
DASM DEBUG - MINIO SIGNED URL EXPIRE MINUTES : 60
DASM DEBUG - nah that didnt work
Error is : InvalidArgumentError: message: Invalid access_key and secret_key.
Internal Server Error: /v2/imagefile/1

This takes me up top the minio API, since that is a method of the Object MINIO i think no? defiined here:
https://docs.min.io/docs/python-client-api-reference.html#presigned_get_object

and that got me back to a simple "is the instantiation of minioExternal Client working", so i wanted to make sure (outside of sreg) that it was working.

So here is my minio_client stub and when we run it, you can see that using that URL as defined and the creds baked in - the client connects and i see buckets. Thus i think that our minioClient connection is fine - i think that our issue is somewhere in the ACCESS/SECRET key passing to the objects.

--- start code ----
#!/bin/python3
from minio import Minio
from minio.error import ResponseError
from datetime import date
import random
import string

## HELPER
def get_random_string(length):
    letters = string.ascii_lowercase
    result_str = ''.join(random.choice(letters) for i in range(length))
    return result_str

minioClient = Minio('singregistry.hosting.cerence.net:9000',
                  access_key='zzzzzz',
                  secret_key='zzzzzzz',
                  secure=True)

print("HELLO MINIO!")
buckets = minioClient.list_buckets()
for bucket in buckets:
   print(bucket.name, bucket.creation_date)

for i in range(5):
   try:
     name = 'make-bucket-' + get_random_string(5)
     print("making bucket : " + name)
     minioClient.make_bucket(name, location="us-east-1")
   except ResposneError as err:
     print(err)

print("Lets see our new list of buckets")
buckets = minioClient.list_buckets()
for bucket in buckets:
   print(bucket.name, bucket.creation_date)

----- END CODE ---

--- OUTPUT FROM CODE ---
[root@ip-0AB36F04 ~]# ./minio_client.py
Today's date: 2020-07-30
HELLO MINIO!
bucket2 2020-07-30 16:42:47.323000+00:00
bucket3-forshow 2020-07-30 17:18:53.203000+00:00
bucket3-show 2020-07-30 17:18:58.322000+00:00
bucket5-show 2020-07-30 17:19:05.885000+00:00
make-bucket-amull 2020-07-30 23:18:59.590000+00:00
make-bucket-eojtf 2020-07-30 23:18:59.566000+00:00
make-bucket-pfqey 2020-07-30 23:18:59.535000+00:00
make-bucket-rrczm 2020-07-30 23:18:59.613000+00:00
make-bucket-zwfik 2020-07-30 23:18:59.638000+00:00
making bucket : make-bucket-idfjv
making bucket : make-bucket-joisv
making bucket : make-bucket-auwis
making bucket : make-bucket-ysbco
making bucket : make-bucket-acxpv
Lets see our new list of buckets
bucket2 2020-07-30 16:42:47.323000+00:00
bucket3-forshow 2020-07-30 17:18:53.203000+00:00
bucket3-show 2020-07-30 17:18:58.322000+00:00
bucket5-show 2020-07-30 17:19:05.885000+00:00
make-bucket-acxpv 2020-07-30 23:19:56.631000+00:00
make-bucket-amull 2020-07-30 23:18:59.590000+00:00
make-bucket-auwis 2020-07-30 23:19:56.586000+00:00
make-bucket-eojtf 2020-07-30 23:18:59.566000+00:00
make-bucket-idfjv 2020-07-30 23:19:56.532000+00:00
make-bucket-joisv 2020-07-30 23:19:56.562000+00:00
make-bucket-pfqey 2020-07-30 23:18:59.535000+00:00
make-bucket-rrczm 2020-07-30 23:18:59.613000+00:00
make-bucket-ysbco 2020-07-30 23:19:56.606000+00:00
make-bucket-zwfik 2020-07-30 23:18:59.638000+00:00

--- END OUTPUT ---


At this point, im going to understand why in the SREG views/images.py the call to get 

url = minioExternalClient.presigned_get_object(bucket_name, object_name, expires=timedelta(days=7))

if failing for you - you can see in the bad bad debug hacks, that (if i read the APi correctly below),

this "should work" in getting the url back ,since we can see from my debug that bucket_name is defined and viable, object_name (which you have as storage in your call passing is populated with the URL you geneated) and timedelta (MINIO_SIGNED_URL_MINUTES) is set to 60, so im digging more to understand that - im not sure i followed the presigned_v4 call you documented (or maybe im not getting there and you see somethign i dont?)>




https://docs.min.io/docs/python-client-api-reference.html#presigned_get_object

**Comment by lmcdasm** [comment]
And finally, when i hardcoded the creds into the MiniCall, it "worked"


DEBUG   [U=0,P=102647]     apiGet()                      apiGet calling v1/containers/daniel.smith/dasm-test/something
DEBUG   [U=0,P=102647]     apiGet()                      apiGet calling v1/images/daniel.smith/dasm-test/something:sha256.b226c379e19853f16f1dc76b13b70deaf822afd6078c5bc3cd9b921d32cf7101?arch=amd64
DEBUG   [U=0,P=102647]     postFileWrapper()             Now uploading to the library
DEBUG   [U=0,P=102647]     postFileV2()                  Using legacy (single part) uploader
DEBUG   [U=0,P=102647]     legacyPostFileV2()            legacyPostFileV2 calling v2/imagefile/3
DEBUG   [U=0,P=102647]     apiCreate()                   apiCreate calling v2/imagefile/3
2.6MiB / 2.6MiB [==============================================================] 100 % 8.8 MiB/s 0s
2.6MiB / 2.6MiB [==============================================================] 100 % 8.8 MiB/s 0s
DEBUG   [U=0,P=102647]     postFileWrapper()             Upload completed OK
DEBUG   [U=0,P=102647]     UploadImage()                 Setting tags against uploaded image
DEBUG   [U=0,P=102647]     UploadImage()                 This library does not support multiple architectures per tag.
DEBUG   [U=0,P=102647]     UploadImage()                 This tag will replace any already uploaded with the same name.
DEBUG   [U=0,P=102647]     getTags()                     getTags calling v1/tags/1
DEBUG   [U=0,P=102647]     setTags()                     Setting tag latest
DEBUG   [U=0,P=102647]     setTag()                      setTag calling v1/tags/1


DASM DEBUG - MINIO SIGNED URL EXPIRE MINUTES : 60
DASM - data :{'uploadURL': 'https://singregistry.hosting.cerence.net:9000/bucket2/dasm-test/something%3Asha256.b226c379e19853f16f1dc76b13b70deaf822afd6078c5bc3cd9b921d32cf7101?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=Sp3ci8l34%2F20200730%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20200730T233424Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=fdfca28d0ee53eb408eb4a93f84ed60c19b709781e45b90b031afb360c369dca'}
[pid: 59|app: 0|req: 6/8] 10.179.111.4 () {36 vars in 497 bytes} [Thu Jul 30 18:34:24 2020] POST /v2/imagefile/3 => generated 428 bytes in 25 msecs (HTTP/1.1 200) 5 headers in 137 bytes (1 switches on core 2)
PUT CompletePushImageFileView
DUMMY-055e367c-9a73-4128-8fd7-cb82490a5f44
[pid: 55|app: 0|req: 2/9] 10.179.111.4 () {36 vars in 512 bytes} [Thu Jul 30 18:34:24 2020] PUT /v2/imagefile/3/_complete => generated 0 bytes in 22 msecs (HTTP/1.1 200) 4 headers in 102 bytes (1 switches on core 0)
[pid: 59|app: 0|req: 7/10] 10.179.111.4 () {34 vars in 453 bytes} [Thu Jul 30 18:34:24 2020] GET /version => generated 58 bytes in 2 msecs (HTTP/1.1 200) 5 headers in 141 bytes (1 switches on core 3)
GET CollectionTagsView

**Comment by vsoch** [comment]
> so i did a bit more debugging and above you mentioned you are getting the "url" fromsregistry_presign_v4 call.. however when i trace to the invoking object, i see its a call to presigned_put_object ( put some try/excepts to play around with my stub)

Yes it does that first to generate the start of the URL, and then customizes it (take a look at minio.py I linked previously)

> And finally, when i hardcoded the creds into the minioClient and minioExternalClient instantiation (in views/minio.py, it "worked")

Great! if you found a solution please open a PR to the repository, and I can test for the http client.

**Comment by lmcdasm** [comment]
Hey there.

I think we should do the following:

a) we can close this ticket in the minio project, since it doesnt "seem" to be an issue there
b) im in the process of figuring out where in the different between the sourching of .minio-env (which has the right creds) and the instantion they are getting lost/not read - or something is going on and if i find the issue there, then will PR on sregistry project with the fix.

i am about to do a " clean install" with this rig on a "fresh node" to make sure i can reproduce (and then we can publish the configuration and setup )..


i feel bad for "brain dumping in this project" for something that i suspect is a buried / sticky file in my env somewhere in the end (im sure ill find it out)  - thanks all for watching :)

**Comment by vsoch** [comment]
But that's still not the multipart... but maybe it's one bug fixed :) 
```
DEBUG [U=0,P=102647] postFileV2() Using legacy (single part) uploader
DEBUG [U=0,P=102647] legacyPostFileV2() legacyPostFileV2 calling v2/imagefile/3
```

**Comment by lmcdasm** [comment]
hehe.. one sec, ill do a big image and try

**Comment by vsoch** [comment]
No worries @lmcdasm ! Let's pick up in sregistry. Thanks for listening to our back and forth minio maintainers! Nothing to see here... lol
